_drafts/device-logging.md

```diff
@@ -0,0 +1,273 @@
+
+At past companies, we developed command line tools that would read log files and print out a text summary of all the exceptions, any network connectivity issues detected, and various other useful information. The benefit of this was massive — the tool could analyze logs much faster than we could individually and in much less time, we could share the tool with our colleagues and save time across the board, we could iterate on adding new features to the tool as the team grew and new scenarios were discovered.
+
+To facilitate that more easily, it would’ve helped if we could rely on the format of logs to be much more machine readable.
+
+#### Avoid multiline logs
+
+Multiline logs are a particularly painful problem for computers to parse. Most system expect logs to all be on a single line, and tracking logs across multiple lines requires retaining state specific to that error message just to parse that error.
+
+```markdown
+E bort    : Network error: -55
+E bort    :     RSSI: 61
+E bort    :     SSID: "Home-Guest"
+```
+
+On one line, it’s slightly less visually readable, but vastly more easily machine parsable.
+
+```markdown
+E bort    : Network error: -55, RSSI: 61, SSID: "Home-Guest"
+```
+
+#### Consider using Structured logs
+
+We can take another page out of server logging best practices [^structuredlogging], and use structured logging. Logs by default are unstructured strings. If we logged more often in a structured format that was easily machine parsable, we could greatly simplify parsing and organizing the data. This is especially useful when we want to eventually parse specific information in the log, otherwise we'd have to write a Regex to parse the log.
+
+```markdown
+E bort    : {"Network error": "-55", "RSSI": 61, "SSID": "Home-Guest"}
+```
+
+JSON is the most common format for structured logs, and can be written manually in code fairly quickly. It's also fairly readable compared to the unstructured log, too.
+
+#### Skip the reading and build visualizations
+
+The end state of logging is ideally to never read the raw logs ourselves. Once we can empower computers to read all sorts of logs, we can build the tools to parse and then visualize the information, and save ourselves the effort of piecing together all the information in our head.
+
+<p align="center">
+  <img width="850" src="{% img_url blog-device-logging/logging-fidelity.png %}" alt="Wi-Fi connection status over time" />
+</p>
+
+If this sounds familiar, this is pitch for Memfault! Instead of reading logs, we can acquire a lot of the same data that the log would have produced, and then display it on a browsable timeline in extremely high resolution.
```

Comment:
Could you add some more details on how we got to this image? Are these metrics? 😉